src/codec.rs

```diff
@@ -885,9 +933,118 @@ impl<T: Encode, const N: usize> Encode for [T; N] {
+		Self::decode_into(DecodeContext::new(), input, &mut array)?;
+
+		// SAFETY: `decode_into` succeeded, so the array is initialized.
+		unsafe {
+			Ok(array.assume_init())
+		}
+	}
+
+    fn decode_into<I: Input>(ctx: DecodeContext, input: &mut I, dst: &mut MaybeUninit<Self>) -> Result<DecodeFinished, Error> {
+		let is_primitive = match <T as Decode>::TYPE_INFO {
+			| TypeInfo::U8
+			| TypeInfo::I8
+				=> true,
+			| TypeInfo::U16
+			| TypeInfo::I16
+			| TypeInfo::U32
+			| TypeInfo::I32
+			| TypeInfo::U64
+			| TypeInfo::I64
+			| TypeInfo::U128
+			| TypeInfo::I128
+			| TypeInfo::F32
+			| TypeInfo::F64
+				=> cfg!(target_endian = "little"),
+			TypeInfo::Unknown => false
+		};
+
+		if is_primitive {
+			// Let's read the array in bulk as that's going to be a lot
+			// faster than just reading each element one-by-one.
+
+			let ptr: *mut [T; N] = dst.as_mut_ptr();
+			let ptr: *mut u8 = ptr.cast();
+			let bytesize = core::mem::size_of::<T>().checked_mul(N).expect("array lengths are sane");
+
+			// TODO: This is potentially slow; it'd be better if `Input` supported
+			//       reading directly into uninitialized memoy.
+			//
+			// SAFETY: The pointer is valid and points to a memory `bytesize` bytes big.
+			unsafe {
+				ptr.write_bytes(0, bytesize);
+			}
+
+			// SAFETY: We've zero-initialized everything so creating a slice here is safe.
+			let slice: &mut [u8] = unsafe {
+				core::slice::from_raw_parts_mut(ptr, bytesize)
+			};
+
+			input.read(slice)?;
+
+			// SAFETY: We've initialized the whole slice so calling this is safe.
+			unsafe {
+				return Ok(ctx.assert_decoding_finished());
+			}
+		}
+
+		let slice: &mut [MaybeUninit<T>; N] = {
+			let ptr: *mut [T; N] = dst.as_mut_ptr();
+			let ptr: *mut [MaybeUninit<T>; N] = ptr.cast();
+			// SAFETY: Casting `&mut MaybeUninit<[T; N]>` into `&mut [MaybeUninit<T>; N]` is safe.
+			unsafe { &mut *ptr }
+		};
+
+		/// A wrapper type to make sure the partially read elements are always
+		/// dropped in case an error occurs or the underlying `decode` implementation panics.
+		struct State<'a, T, const N: usize> {
+			count: usize,
+			slice: &'a mut [MaybeUninit<T>; N]
+		}
+
+		impl<'a, T, const N: usize> Drop for State<'a, T, N> {
+			fn drop(&mut self) {
+				if !core::mem::needs_drop::<T>() {
+					// If the types don't actually need to be dropped then don't even
+					// try to run the loop below.
+					//
+					// Most likely won't make a difference in release mode, but will
+					// make a difference in debug mode.
+					return;
+				}
+
+				for item in &mut self.slice[..self.count] {
+					// SAFETY: Each time we've read a new element we incremented `count`,
+					//         and we only drop at most `count` elements here,
+					//         so all of the elements we drop here are valid.
+					unsafe {
+						item.assume_init_drop();
+					}
+				}
+			}
+		}
+
+		let mut state = State {
+			count: 0,
+			slice
+		};
+
+		while state.count < state.slice.len() {
```

Comment:
How about `for` loop to enable possible vectorization?

Reply:
state.slice[state.count] Inserts a bound check which can be eliminated with for and the cycle can be optimized for primitives like u32. Just a guess :D

Reply:
AFAIK the compiler should be able to elide the bounds check here as-is. Whether the loop is written with a `for` or with a `while` doesn't matter.

Nevertheless, it's unlikely that there will an opportunity for vectorization here though. Everything inside the loop would have to be inlined, the decoded `T` would have to be `Copy`, and reading from the `input` would also need to have its bounds elided.

One possible extra optimization that could be done here would be to calculate the minimum amount of bytes that has to be read to decode one `T`, multiply that by the number of elements, and then add a new method to `Input` which would use that size to insert a precheck (if possible for a given `Input`) that would allow the compiler to elide the bounds when reading from `Input`. But that's out of scope for this PR.